### DISCUSSION TITLE
Horizontal Stacked Bar Chart Timeline

### DISCUSSION THREAD

**Discussion by scottfranz** [observablehq/plot#2176] [Q&A]
Hello,

I am trying to create a horizontal stacked bar chart that shows survey administrations of a longitudinal panel study. I would like each bar to represent each cohort. Then I would like to color a portion of each bar based on how many follow-up surveys they should have completed (how far along in the study they are).

I am using this observable example to visualize each panel cohort. [Plot: Civilizations timeline / Observable | Observable](https://observablehq.com/@observablehq/plot-civilizations-timeline)
So the X-axis is time (or age) and the Y-axis is the cohort. But I would like to color portions of each bar based on how many follow up surveys they completed. Similar to something like this example [Plot: Horizontal stacked bars / Observable | Observable](https://observablehq.com/@observablehq/plot-horizontal-stacked-bars)

Except instead of grouping “counts” of a categorical variable like party affiliation. I would like to color portions of the bars based on ranges in time where they should have filled out a follow up surveys. Because I am fabricating this data and do not have a bunch of rows to count. Is this possible am I on the wrong track with the bar mark? I have not created a variable to use to color the bars yet but essentially I want the first 10 years to be a color, then the next 10 years to be a different color to denote which follow-up period they are in. Here is the code I have so far.
````
mtf = { 
  const subset = data.map(
    ({ Panel, Start, End, AgeStart, AgeEnd }) => ({
      Panel: Panel,
      Start: parser(Start),
      End: parser(End),
      AgeStart: AgeStart,
      AgeEnd: AgeEnd
    })
  );

  return subset;
}

formatter = d3.timeFormat("%Y")

Plot.plot({
  marginLeft: 130,
  axis: null,
  x: {
    axis: "top",
    grid: true,
  },
  marks: [
    Plot.barX(mtf, {
      x1: "Start",
      x2: "End",
      y: "Panel",
      sort: {y: "x1"}
    }),
    Plot.text(mtf, {
      x: "Start",
      y: "Panel",
      text: (d) => `${formatter(d.Start)}`,
      textAnchor: "end",
      dx: -3
    })
  ]
})
````

Here is the csv file I created [mtfdata.csv](https://github.com/user-attachments/files/17116434/mtfdata.csv). Also here is the bar chart I just need to figure out how to color portions of the bars.
![bar chart](https://github.com/user-attachments/assets/60b540f9-fb62-4a74-a8f7-55c5e467a877)

**Comment by scottfranz** [comment]
I was also able to formulate the data in a different way to try to color the bars. Here is the new data
[panel.csv](https://github.com/user-attachments/files/17116980/panel.csv). Here is the code:

````
Plot.plot({
  //x: { tickFormat: Math.abs, label: "# of answers" },
  y: { tickSize: 0 },
  color: {
    legend: true,
    domain: [
      "12th Grade",
      "1st Follow Up",
      "2nd Follow Up",
      "3rd Follow Up",
      "4th Follow Up",
      "5th Follow Up",
      "6th Follow Up",
      "7th Follow Up",
      "8th Follow Up",
      "9th Follow Up",
      "10th Follow Up",
      "11th Follow Up",
      "12th Follow Up",
      "13th Follow Up",
    ],
    scheme: "blues"
  },
  marks: [
    Plot.barX(
      panel,
      Plot.groupY(
        { x: "count" },
        {
          fill: "Age Group",
          order: [
            "12th Grade",
            "1st Follow Up",
            "2nd Follow Up",
            "3rd Follow Up",
            "4th Follow Up",
            "5th Follow Up",
            "6th Follow Up",
            "7th Follow Up",
            "8th Follow Up",
            "9th Follow Up",
            "10th Follow Up",
            "11th Follow Up",
            "12th Follow Up",
            "13th Follow Up",
          ],
          y: "Panel"
        }
      )
    )
  ]
})
````
Here is the output:
![stacked bar charts](https://github.com/user-attachments/assets/cf0c223e-c701-470c-a7a9-5198f7c3d7e0)

Instead of the frequency of counts I would like the x-axis to be years.

**Reply by scottfranz** [reply]
I am getting closer to what I want. I figured out a better way to organize the data [mtfdata3.csv](https://github.com/user-attachments/files/17133401/mtfdata3.csv). Now I am having trouble with both sorting on the "x1" channel and with my text labels. For some reason the sorting is different around "Cohort 21". Also how do I just show the first text label but not the others? Thanks!

````
Plot.plot({
  marginLeft: 130,
  axis: null,
  color: {
     legend: true
  },
  x: {
    axis: "top",
    grid: true,
  },
  marks: [
    Plot.barX(mtf, {
      x1: "Start",
      x2: "End",
      y: "Cohort",
      fill: "Group",
      sort: {y: "x1"}
    }),
    Plot.text(mtf, {
      x: "Start",
      y: "Cohort",
      text: "Cohort",
      textAnchor: "end",
      dx: -3
    })
  ]
})
````
![bar chart colors](https://github.com/user-attachments/assets/4838f841-3bfd-499e-8e57-ad0030b175cd)

**Reply by scottfranz** [reply]
Something about all Cohorts above 32 having not having a 5-Year Follow Up value. I am confused about the default sorting behavior as well. Here is a picture.
![default order](https://github.com/user-attachments/assets/3b630979-d762-4911-91e6-d2fa53efc9b1)

**Reply by scottfranz** [reply]
Almost have it where I want it. I had to manually set the domain on the y axis. Now I am hoping to add text labels on the right side that show the age.
![almost finished bar chart](https://github.com/user-attachments/assets/be8aafe8-5947-4343-8490-9f839bce7a9c)

**Reply by Fil** [reply]
the default reducer for the sort option is "max"; maybe you want to try "min" instead:

`sort: {y: "x1", reduce: "min"}`

or in long form:

`sort: {y: {value: "x1", reduce: "min"}}`

**Reply by scottfranz** [reply]
Thanks that helped!
